src/dmd/dinifile.d

```diff
@@ -144,20 +144,13 @@ void updateRealEnvironment(StringTable* environment)
 {
     static int envput(const(StringValue)* sv)
     {
-        const name = sv.toDchars();
-        const namelen = strlen(name);
-        const value = cast(const(char)*)sv.ptrvalue;
+        const name = sv.toString();
+        const value = (cast(const(char)*)sv.ptrvalue)[0 .. sv.length];
         if (!value) // deleted?
             return 0;
-        const valuelen = strlen(value);
-        auto s = cast(char*)malloc(namelen + 1 + valuelen + 1);
-        assert(s);
-        memcpy(s, name, namelen);
-        s[namelen] = '=';
-        memcpy(s + namelen + 1, value, valuelen);
-        s[namelen + 1 + valuelen] = 0;
+        const s = name ~ '=' ~ value ~ '\0';
         //printf("envput('%s')\n", s);
-        putenv(s);
+        putenv(cast(char*)s.ptr);
```

Comment:
That cast shouldn't be necessary

Reply:
Although the windows version of `putenv()` takes a const char* parameter, the POSIX flavour does not.

Reply:
And that should tip you off as a good reason not to `cast`:

`man putenv` on my system (Mac OS) says:
> The string pointed to by string becomes part of the environment.  A program should not alter or free the string, and should not use stack or other transient string variables as arguments to putenv().  The setenv() function is strongly preferred to putenv().

Since the string has no reference to it kept anywhere, the GC is free to collect it, which would lead to random memory corruption.